pennylane/tape/tapes/jacobian_tape.py

```diff
@@ -438,34 +477,62 @@ def jacobian(self, device, params=None, **options):
                 # this computation is only performed once, for all parameters.
                 options["y0"] = np.asarray(self.execute_device(params, device))
 
-        jac = None
-        p_ind = range(len(params))
+        options["device"] = device
+
+        # collect all circuits (tapes) and postprocessing functions required
+        # to compute the jacobian
 
-        # loop through each parameter and compute the gradient
-        for idx, (l, param_method) in enumerate(zip(p_ind, allowed_param_methods)):
+        all_tapes = []
+        reshape_info = []
+        processing_fns = []
 
-            if param_method == "0":
-                # Independent parameter. Skip, as this parameter has a gradient of 0.
-                continue
+        for trainable_idx, param_method in enumerate(diff_methods):
 
             if (method == "best" and param_method[0] == "F") or (method == "numeric"):
                 # finite difference method
-                g = self.numeric_pd(l, device, params=params, **options)
+                tapes, processing_fn = self.numeric_pd(trainable_idx, params=params, **options)
 
             elif (method == "best" and param_method[0] == "A") or (method == "analytic"):
+                # some analytic methods need the device wires
+                options["dev_wires"] = device.wires
+
                 # analytic method
-                g = self.analytic_pd(l, device, params=params, **options)
+                tapes, processing_fn = self.analytic_pd(trainable_idx, params=params, **options)
+
+            # we create a flat list here to feed at once to the device
+            all_tapes.extend(tapes)
+            # to extract the correct result for this parameter later, remember the number of tapes
+            reshape_info.append(len(tapes))
+            processing_fns.append(processing_fn)
+
+        # execute all tapes
+        results = [
+            tape.execute(device) for tape in all_tapes
+        ]  # soon to be: device.batch_execute(all_tapes)
+
+        if method == "numeric":
+            output_dim = all_tapes[0]._output_dim
+
+            if self.output_dim != output_dim:
+                self._output_dim = output_dim
```

Comment:
Maybe a comment for this one might be helpful :thinking:

Reply:
Oh, this needs to be removed. It is a hotfix to get some of the tests working.

Reply:
Uh good point. I think we are just trying to infer the output dimension (influencing the size of the jacobian) from the first tape in `tapes`. This can fail, for example if the tapes needed to evaluate gradients have different outputs that the original tape. @josh, what was the conduct here? Is this watertight? Can't we always run `y0` in `jacobian` and infer the dimension from that?

Reply:
I've removed this in the PR that I'm about to make!